src/relo_core.c

```diff
@@ -1472,39 +1485,116 @@ struct btf *bpf_reloc_info_get_btf(struct btf_reloc_info *info) {
+
+	type_id = spec->root_type_id;
+	t = btf__type_by_id(spec->btf, type_id);
+	s = btf__name_by_offset(spec->btf, t->name_off);
+
+	printf("RELOCATION: [%u] %s %s", type_id, btf_kind_str(t), str_is_empty(s) ? "<anon>" : s);
+
+	if (core_relo_is_type_based(spec->relo_kind)) {
+		printf("\n");
+		return;
+	}
+
+	if (core_relo_is_enumval_based(spec->relo_kind)) {
+		t = skip_mods_and_typedefs(spec->btf, type_id, NULL);
+		e = btf_enum(t) + spec->raw_spec[0];
+		s = btf__name_by_offset(spec->btf, e->name_off);
+
+		printf("::%s = %u\n", s, e->val);
+		return;
+	}
+
+	if (core_relo_is_field_based(spec->relo_kind)) {
+		for (i = 0; i < spec->len; i++) {
+			if (spec->spec[i].name)
+				printf(".%s", spec->spec[i].name);
+			else if (i > 0 || spec->spec[i].idx > 0)
+				printf("[%u]", spec->spec[i].idx);
+		}
+
+		printf(" (");
+		for (i = 0; i < spec->raw_len; i++)
+			printf("%s%d", i == 0 ? "" : ":", spec->raw_spec[i]);
+
+		if (spec->bit_offset % 8)
+			printf(" @ offset %u.%u)", spec->bit_offset / 8, spec->bit_offset % 8);
+		else
+			printf(" @ offset %u)", spec->bit_offset / 8);
+
+		printf("\n");
+		return;
+	}
+}
+
+static int btf_reloc_info_gen_field(struct btf_reloc_info *info, struct bpf_core_spec *targ_spec) {
+
 	struct btf *btf = (struct btf *) targ_spec->btf;
 	struct btf_type *btf_type;
 	struct btf_reloc_type *reloc_type;
 
+	btf_reloc_dump_spec(targ_spec);
+
 	btf_type = btf_type_by_id(btf, targ_spec->root_type_id);
 
 	reloc_type = btf_reloc_get_type(info, btf_type);
-
 	if (reloc_type == NULL) {
 		reloc_type = calloc(1, sizeof(struct btf_reloc_type));
-
 		reloc_type->type = btf_type;
 		reloc_type->id = targ_spec->root_type_id;
-
 		btf_reloc_add_type(btf, info, reloc_type);
 	}
 
+	struct btf_array *a;
+	struct btf_type *this;
+
 	// add types for members of parent type (only for struct or union)
 	for (int i = 1; i < targ_spec->raw_len; i++) {
-		if (btf_is_array(btf_type)) {
-			struct btf_array *a = btf_array(btf_type);
-			struct btf_type *array_type = (struct btf_type *)btf__type_by_id(btf, a->type);
-
-			// set the current type to array type and continue processing
-			reloc_type = btf_reloc_get_type(info, array_type);
 
+		if (!targ_spec->spec[i].name)
 			continue;
+		else {
+			while (btf_is_mod(btf_type) || btf_is_typedef(btf_type)) {
+				btf_type = (struct btf_type*) btf__type_by_id(btf, btf_type->type);
+				reloc_type = btf_reloc_get_type(info, btf_type);
+			}
 		}
 
-		if (!btf_is_struct(btf_type) && !btf_is_union(btf_type)) {
-			printf("received a type not expected\n");
+		switch (btf_kind(btf_type)) {
+		case BTF_KIND_STRUCT:
+		case BTF_KIND_UNION:
+			break;
+		case BTF_KIND_ARRAY:
+			a = btf_array(btf_type);
+			this = (struct btf_type *)btf__type_by_id(btf, a->type);
+			reloc_type = btf_reloc_get_type(info, this);
+			continue;
+		//case BTF_KIND_TYPEDEF:
+		//	reloc_type = btf_reloc_get_type(info, btf_type);
+		//	btf_type = (struct btf_type*) btf__type_by_id(btf, btf_type->type);
+		//	continue;
```

Comment:
To make it work you need to:

1. `break` instead of `continue`: After solving the typedef we still want the logic below to be executed to add the members for this type. 
2. `reloc_type = btf_reloc_get_type(info, btf_type);` ->  `reloc_type = btf_reloc_get_type(info, btf_type->type);` We want `reloc_type` to be the anon struct and not the typedef. (too add the members to the anon struct below). 


```suggestion
		//case BTF_KIND_TYPEDEF:
		//	reloc_type = btf_reloc_get_type(info, btf_type->type);
		//	btf_type = (struct btf_type*) btf__type_by_id(btf, btf_type->type);
		//	break;
```

Reply:
Check my previous answer, I think it covers this case. It wont work here because we need the full typedef/mods resolution, like the **skip_mods_or_typedefs** function (but considering reloc re-definition for our logic).

Reply:
yeah, you're right. It's better to skip them at the beginning.